### PR TITLE
Add a configuration option to allow enable/disable writing error log to ZK

### LINKED ISSUE #1486: Improve Error log to Zookeeper server
Currently Helix (Participant) sends error logging to ZK and Helix controller would aggregate and remove these logs.

Some failing path of this process can result in repeated sending error messages to ZK, which brings down ZK.

For example, HelixTaskExecutor.OnMessage(), if failing to create messageHandler, would result in logging error to ZK and remove messages. Next Controller would send another messages and repeat this process. So on and so forth.

We should give an option to enable logging and by default disable these loggings, as Zookeeper is meta data store, should not be used as database or datastore.

### PR DESCRIPTION
**Pull request by kaisun2000**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

    fix #1486

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

    Improve statusUpdateUtil log error to ZK by adding an option to enabled
    it. By default, it would not log error to ZK. This is to avoid some
    error code path that keep flooding ZK sever which cause DoS to Zk.
    Such as HelixTaskExecutor onMessage creation messageHandler exception.

### Tests

- [x] The following tests are written for this issue:
  TestStatusUpdateUtil

- [x] The following is the result of the "mvn test" command on the appropriate module:

github passed.
https://github.com/apache/helix/runs/1344486165?check_suite_focus=true

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-common/src/main/java/org/apache/helix/SystemPropertyKeys.java
Comment (lei-xia) [inline helix-common/src/main/java/org/apache/helix/SystemPropertyKeys.java:86]: minor: "helix.StateUpdateUtil.errorLogToZK.enabled"?  We still log error to local logfile, right?

File: helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java
Comment (lei-xia) [inline helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java:62]: Please use HelixUtil.getSystemPropertyAsBoolean(), it allows you to specify the default value there in the method.

File: helix-core/src/test/java/org/apache/helix/integration/manager/TestParticipantManager.java
Comment (lei-xia) [inline helix-core/src/test/java/org/apache/helix/integration/manager/TestParticipantManager.java:70]: Same here, use HelixUtil.

File: helix-common/src/main/java/org/apache/helix/SystemPropertyKeys.java
Comment (kaisun2000) [reply helix-common/src/main/java/org/apache/helix/SystemPropertyKeys.java:86]: This is not minor, this is a MAJOR, a very good point. Let me make sure we have error level log out.

File: helix-core/src/test/java/org/apache/helix/integration/manager/TestParticipantManager.java
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/integration/manager/TestParticipantManager.java:70]: In this test, here we are setting this property so that later test which asserting existence of error record in Zk won't fail. HelixUtil does not apply here as it is used to retrieve record.

File: helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java
Comment (kaisun2000) [reply helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java:62]: There is only HelixUtil.getSystemPropertyAsInt/Long, no asBoolean(). Let me add as Boolean then, make the changes a little bit wider.

File: helix-core/src/test/java/org/apache/helix/util/TestStatusUpdateUtil.java
Comment (lei-xia) [inline helix-core/src/test/java/org/apache/helix/util/TestStatusUpdateUtil.java:24]: Can you test enable it by the new config?

Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/util/TestStatusUpdateUtil.java:24]: It is already tested in TestParticipantManager. testSessionExpiryInTransition assert the error log znode is not null. Do we need to add another explicit one here?

File: helix-core/src/main/java/org/apache/helix/util/HelixUtil.java
Comment (pkuwm) [inline helix-core/src/main/java/org/apache/helix/util/HelixUtil.java:65]: This import is not used.

File: helix-core/src/main/java/org/apache/helix/util/HelixUtil.java
Comment (pkuwm) [inline helix-core/src/main/java/org/apache/helix/util/HelixUtil.java:479]: We don't have to create this method to be redundant.
Simply, `Boolean.getBoolean(propertyKey, propertyDefaultValue)` is already what is needed, right?

**Comment by lei-xia** [comment]
The PR title seems not grammar correct, maybe "Add a configuration option to allow enable/disable writing error log to ZK"?

File: helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java
Comment (pkuwm) [inline helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java:60]: `Boolean.getBoolean(propertyKey, propertyDefaultValue)` is simple and good enough?

File: helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java
Comment (pkuwm) [inline helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java:568]: Nit, `record.toString()` is redundant, `record` is enough as `toString()` would be auto called.

File: helix-core/src/test/java/org/apache/helix/integration/manager/TestParticipantManager.java
Comment (pkuwm) [inline helix-core/src/test/java/org/apache/helix/integration/manager/TestParticipantManager.java:70]: Can you help me understand why you set the property here, but clear it in `TestStatusUpdateUtil`? Why do we not put both in the test method? Otherwise, the system config would affect other tests, right?

File: helix-core/src/test/java/org/apache/helix/util/TestStatusUpdateUtil.java
Comment (pkuwm) [inline helix-core/src/test/java/org/apache/helix/util/TestStatusUpdateUtil.java:2]: Missing apache license. Maybe you could setup apache license in your intellij?

**Comment by kaisun2000** [comment]
> The PR title seems not grammar correct, maybe "Add a configuration option to allow enable/disable writing error log to ZK"?

changed.

File: helix-core/src/test/java/org/apache/helix/util/TestStatusUpdateUtil.java
Comment (pkuwm) [inline helix-core/src/test/java/org/apache/helix/util/TestStatusUpdateUtil.java:69]: Why `assertTrue(True);` which is always true? Just put a comment it is expected. Or to be more accurate, we can assert the exception message to ensure the exception is the expected one not others(same exception type but different message thrown from other places).

File: helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java:569]: IMHO,

if (!ERROR_LOG_TO_ZK_ENABLED) {
  _logger.error("StatusUpdate Error record: {}", record.toString());
  return;
}
...

File: helix-core/src/test/java/org/apache/helix/util/TestStatusUpdateUtil.java
Comment (jiajunwang) [inline helix-core/src/test/java/org/apache/helix/util/TestStatusUpdateUtil.java:73]: This is a potential thread leak if test failure.

File: helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java
Comment (lei-xia) [reply helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java:569]: Not actually, even we disable writting to ZK, we still need to log into local log file, right?

File: helix-common/src/main/java/org/apache/helix/SystemPropertyKeys.java
Comment (lei-xia) [inline helix-common/src/main/java/org/apache/helix/SystemPropertyKeys.java:86]: Again, we are disabling log to ZK, not disabling error log (to log file) at all, right?

File: helix-core/src/test/java/org/apache/helix/util/TestStatusUpdateUtil.java
Comment (lei-xia) [reply helix-core/src/test/java/org/apache/helix/util/TestStatusUpdateUtil.java:24]: It is better to add both positive and negative tests in here. The main purpose of the other test is not testing error log though.

File: helix-common/src/main/java/org/apache/helix/SystemPropertyKeys.java
Comment (dasahcc) [reply helix-common/src/main/java/org/apache/helix/SystemPropertyKeys.java:86]: We have a class ZkSystemPropertyKeys to hold this kind of system property. Please put this there.

File: helix-core/src/main/java/org/apache/helix/util/HelixUtil.java
Comment (dasahcc) [reply helix-core/src/main/java/org/apache/helix/util/HelixUtil.java:479]: +1. Unless you need some error handling in the function.

File: helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java:569]: I don't have a strong preference here.
But I think it depends on how we treat this util, is it a more powerful logger? Or it is a status updater like the name suggested.

The current implementation is treating this class as a pure status updater. There is no extra log done unless the update process fails. And I think it is clean. If a log is needed, then the caller shall call logger method in parallel with the publishErrorRecord() method.

On the other hand, if we want to twist this class to be a more powerful logger, then we will have mixed usage. Only part of the code is using it. And whether to use it or not is not clear.

Overall, I treat the log here as a fallback logic when ERROR_LOG_TO_ZK_ENABLED == false. So it shall not log when the util still writes to ZK.

Comment (kaisun2000) [reply helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java:569]: So let us keep it this way. The reasoning is that source of truth logging is in log file.

File: helix-common/src/main/java/org/apache/helix/SystemPropertyKeys.java
Comment (kaisun2000) [reply helix-common/src/main/java/org/apache/helix/SystemPropertyKeys.java:86]: @lei-xia, so let us change the name to " STATEUPDATEUTIL_ERROR_LOG2ZK_ENABLED?

Comment (kaisun2000) [reply helix-common/src/main/java/org/apache/helix/SystemPropertyKeys.java:86]: @dasahcc, ZkSystemPropertyKeys is for Zookeeper related. This is more to me like logging related.

File: helix-core/src/main/java/org/apache/helix/util/HelixUtil.java
Comment (kaisun2000) [reply helix-core/src/main/java/org/apache/helix/util/HelixUtil.java:65]: removed.

File: helix-core/src/main/java/org/apache/helix/util/HelixUtil.java
Comment (kaisun2000) [reply helix-core/src/main/java/org/apache/helix/util/HelixUtil.java:479]: removed.

File: helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java
Comment (kaisun2000) [reply helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java:568]: changed.

File: helix-core/src/test/java/org/apache/helix/util/TestStatusUpdateUtil.java
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/util/TestStatusUpdateUtil.java:2]: added.

File: helix-core/src/test/java/org/apache/helix/util/TestStatusUpdateUtil.java
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/util/TestStatusUpdateUtil.java:73]: fixed in tearing down in afterClass()

File: helix-core/src/test/java/org/apache/helix/integration/manager/TestParticipantManager.java
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/integration/manager/TestParticipantManager.java:70]:
```
  public void testSessionExpiryInTransition() throws Exception {

    String errPath = PropertyPathBuilder.instanceError(clusterName, "localhost_12918", oldSessionId,
        "TestDB0", "TestDB0_0");
    ZNRecord error = _gZkClient.readData(errPath);
    Assert.assertNotNull(error,
        "InterruptedException should happen in old session since task is being cancelled during handleNewSession");

```

To make sure the following test would work.

File: helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java:569]: I'm fine with both designs. But as I mentioned, this change is half-way. Please either ensure that the source of truth is with the log for ALL the status util call (including log error, log status, etc.) or please don't log here otherwise the util is mixed up with some methods with log and the others don't.

File: helix-core/src/test/java/org/apache/helix/integration/manager/TestParticipantManager.java
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/integration/manager/TestParticipantManager.java:70]: ERROR_LOG_TO_ZK_ENABLED in StatusUpdateUtil is static. I changed to use reflection to control the value now. So there is no need to set or clear it in `TestStatusUpdateUtil` anymore.

File: helix-core/src/test/java/org/apache/helix/util/TestStatusUpdateUtil.java
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/util/TestStatusUpdateUtil.java]: both added now.

File: helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java
Comment (kaisun2000) [reply helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java:568]: Good point. The log are in `publishStatusUpdateRecord' and `publishErrorRecord`.

Here publishStatusUpdateRecord would never log to Zk anymore. The participant side, it has trace level log to log4j, but not controller side. Let me revise it to the way that both controller and participant side has log to log4j then.

**Comment by kaisun2000** [comment]
@dasahcc , @pkuwm , @jiajunwang , ping?

File: helix-common/src/main/java/org/apache/helix/SystemPropertyKeys.java
Comment (jiajunwang) [inline helix-common/src/main/java/org/apache/helix/SystemPropertyKeys.java:86]: I thought more about this configuration. Can we directly make it an enum now instead of changing it later? Otherwise, we will need to deprecate this item and go through the painful backward compatibility checklist in the future.

I suggest making it STATEUPDATE_ERROR_REPORTING_CHANNEL.

File: helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java:504]: nit, we can move this line out of the if-else block too.

File: helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java:522]: same here.

File: helix-common/src/main/java/org/apache/helix/SystemPropertyKeys.java
Comment (kaisun2000) [reply helix-common/src/main/java/org/apache/helix/SystemPropertyKeys.java:86]: Can you illustrate the idea how to make it enum? Something like 

```
enum STATEUPDATE_ERROR_REPORTING_CHANNEL {
    LOG4J,
    LOG2ZK
};

```

and in the system property, 
`public static final String STATEUPDATEUTIL_ERROR_LOG2ZK_ENABLED = <enum value>`;
then we will convert the system property to enum value?

Comment (kaisun2000) [reply helix-common/src/main/java/org/apache/helix/SystemPropertyKeys.java:86]: STATEUPDATEUTIL_ERROR_PERSISTANCY_ENABLED

File: helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java
Comment (pkuwm) [inline helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java:559]: Do we really need this error log for the whole record? It may eat up the disk, imagine if znrecord is close to 1 MB. I am not sure if we really need the record for troubleshooting. Maybe logging the instanceName, updateSubPath, sessionId, etc. is good enough?

Comment (kaisun2000) [reply helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java:559]: This is not a concern. Previously the issue is that N participant would write the log to one Zk quorum at the same time repeatedly. Here, we are writing to N participant at the same time. If this is a concern, we need to revisit all the participant side logging then.

File: helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java
Comment (kaisun2000) [reply helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java:504]: fixed.

File: helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java
Comment (kaisun2000) [reply helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java:522]: fixed.

File: helix-common/src/main/java/org/apache/helix/SystemPropertyKeys.java
Comment (kaisun2000) [reply helix-common/src/main/java/org/apache/helix/SystemPropertyKeys.java:86]: changed.

File: helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java:59]: nit, private?
Since we already doing reflection, I think making it private won't block the test, right?

**Review by jiajunwang — APPROVED** [review]
LGTM! One minor suggestion.

File: helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java
Comment (kaisun2000) [reply helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java:59]: Actually it is not the case. Making it private won't work as I just test. I am not an expert in this area. If you know other trick, let me know.

**Comment by kaisun2000** [comment]
The PR is approved, please help to merge

> Improve statusUpdateUtil log error to ZK by adding an option to enabled
it. By default, it would not log error to ZK. This is to avoid some
error code path that keep flooding ZK sever which cause DoS to Zk.
Such as HelixTaskExecutor onMessage creation messageHandler exception.

File: helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java
Comment (kaisun2000) [reply helix-core/src/main/java/org/apache/helix/util/StatusUpdateUtil.java:59]: This is not the case. Just tested, it does not work. I am not an expert in this area. Feel free to let me know if you know the trick here.

File: helix-core/src/test/java/org/apache/helix/integration/manager/TestParticipantManager.java
Comment (pkuwm) [inline helix-core/src/test/java/org/apache/helix/integration/manager/TestParticipantManager.java:71]: The system property is JVM level. If we don't unset the config, later tests will be using this config. Maybe there are no tests using the config yet. But I suggest, as a good practice, we also unset the config to make unit tests isolated, ensuring other tests won't be impacted by this test's configs.

**Comment by pkuwm** [comment]
I'll merge this PR once the PR CI is successful.

**Pull request merged**
